cmd/kubectl-rbac-flatten/README.md

```diff
@@ -0,0 +1,113 @@
+
+Errors are logged to stderr, and as much processing as possible is
+done.
+
+There are two possible output formats, chosen by a command line flag.
+
+### JSON Output
+
+
+The JSON output does a minimal outer product, producing a tuple per
+(Subject, PolicyRule). The tuples also include some additional fields
+describing their source. The precise data type is as follows, where
+`NamespacedName` is imported from `k8s.io/apimachinery/pkg/types`.
+
+```go
+type Tuple struct {
+	Binding  NamespacedName
+	RoleName string
+	Subject  rbac.Subject
+	Rule     rbac.PolicyRule
+}
+```
+
+The output is one JSON value, an array. Each tuple is written on one
+line, with the comma separators on other lines, making it easy to use
+text-based tools for simple searches. For non-trivial searching, you
+can use a real JSON processor.
+
+Following is an example of JSON output.
+
+```json
+[
+{"Binding":{"Namespace":"","Name":"cluster-admin"},"RoleName":"cluster-admin","Subject":{"kind":"Group","apiGroup":"rbac.authorization.k8s.io","name":"system:masters"},"Rule":{"verbs":["*"],"apiGroups":["*"],"resources":["*"]}}
+,
+{"Binding":{"Namespace":"","Name":"cluster-admin"},"RoleName":"cluster-admin","Subject":{"kind":"Group","apiGroup":"rbac.authorization.k8s.io","name":"system:masters"},"Rule":{"verbs":["*"],"nonResourceURLs":["*"]}}
+,
+...
+]
+```
+
+### Tabular Ouptput
+
+The tabular output does a maximal cross product, so that no tuple in
+the result has a list/slice/array-structured field. The tabular output
+has two tables, one for resource-based rules and one for non-resource
+URLs. The second one is omitted if `--api-goups` is given anything
+other than `*`.
+
+The fields in one row are separated by TABs. The empty string API
+group is output as an empty string (thus between two TABs).
+
+The resource-based table omits the column for the role, for brevity.
+
+Following are excerpts from example tabular output.
+
+```
+BINDING                                                         SUBJECT                                                        VERB               APIGROUP                        RESOURCE                                    OBJNAME
+/cluster-admin                                                  G:system:masters                                               *                  *                               *                                           *
+/ingress-nginx                                                  SA:ingress-nginx/ingress-nginx                                 list                                               configmaps                                  *
+/ingress-nginx                                                  SA:ingress-nginx/ingress-nginx                                 list                                               endpoints                                   *
+...
+
+BINDING                                       ROLE                                      SUBJECT                                          VERB   NRURL
```

Comment:
I think we can safely say ClusterRole for non-resource URLs.

Reply:
True, but I think that would be a confusing irregularity: to say "ClusterRole" in one table and not "Role or ClusterRole" in the other.